### PR TITLE
[3rd tvOS]tvos commands

### PR DESCRIPTION
**Pull request by KazuCocoa**
Final PR for https://github.com/appium/WebDriverAgent/pull/148 as implementing features for tvOS.
As WDA, we can handle tvOS via WDA after this PR.

We need to add tvOS support in xcuitest-driver.

---

```
curl -X GET http://localhost:8100/session/493E964A-FD36-4B34-9AF0-97698B0FE907/wda/element/focused # get a focused element

# After find element. e.g. 17000000-0000-0000-7E6D-000000000000
curl -X POST http://localhost:8100/session/493E964A-FD36-4B34-9AF0-97698B0FE907/element/6B030000-0000-0000-7E6D-000000000000/focuse -d '' # Work via inspector
curl -X GET http://localhost:8100/session/493E964A-FD36-4B34-9AF0-97698B0FE907/element/6B030000-0000-0000-7E6D-000000000000/focused # Work

# For compatibility of clients. Clients can get `focus` status as `get attribute` API
curl -X GET http://localhost:8100/session/493E964A-FD36-4B34-9AF0-97698B0FE907/element/6B030000-0000-0000-7E6D-000000000000/attribute/focused # Work

# Worked. We can click an element on TV.
# Users can do selecting action via `click` as same as iOS
curl -X POST http://localhost:8100/session/493E964A-FD36-4B34-9AF0-97698B0FE907/element/6B030000-0000-0000-7E6D-000000000000/click -d ''
```

---

`hasFocus` was only in tvOS. So I've added ifdef for tvOS again.
I'll add unit tests as next PR since it needs to add new test targets and test classes.

----

# New endpoints

-  `[[FBRoute GET:@"/element/:uuid/attribute/focused"] respondWithTarget:self action:@selector(handleGetFocused:)],`
    - Get an attribute of 'focused'
- `[[FBRoute POST:@"/wda/element/:uuid/focuse"] respondWithTarget:self action:@selector(handleFocuse:)],`
    - Set a `focus`  to an element which is `:uuid`
- `[[FBRoute GET:@"/wda/element/focused"] respondWithTarget:self action:@selector(handleFindFocusedElement:)],`
    - Get an element which is focused

### CODE REVIEW

File: WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m [hunk 1]
```diff
@@ -24,9 +24,9 @@ @implementation XCUIElement (FBTVFocuse)
 
 - (BOOL)fb_setFocusWithError:(NSError**) error
 {
-  [[FBApplication fb_activeApplication] fb_waitUntilSnapshotIsStable];
+  [FBApplication.fb_activeApplication fb_waitUntilSnapshotIsStable];
 
-  if (!self.wdEnabled) {
+  if (!self.isEnabled) {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m]: why not wdEnabled?

File: WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m [hunk 2]
```diff
@@ -36,6 +36,7 @@ - (BOOL)fb_setFocusWithError:(NSError**) error
 
   FBTVNavigationTracker *tracker = [FBTVNavigationTracker trackerWithTargetElement:self];
   for (int i = 0; i < MAX_ITERATIONS_COUNT; i++) {
+    // Here's hasFocus works so far. Maybe, it is because here is handled by `XCUIRemote`...
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m]: Here's -> Here

Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m]: because it is handled

File: WebDriverAgentLib/Commands/FBCustomCommands.m [hunk 3]
```diff
@@ -110,6 +115,13 @@ + (NSArray *)routes
   return FBResponseWithOK();
 }
 
+#pragma mark - Helpers
+
++ (BOOL) isKeyboardPresent {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBCustomCommands.m]: no space before is

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 4]
```diff
@@ -67,7 +67,11 @@ + (NSArray *)routes
     [[FBRoute GET:@"/screenshot/:uuid"] respondWithTarget:self action:@selector(handleElementScreenshot:)],
     [[FBRoute GET:@"/wda/element/:uuid/accessible"] respondWithTarget:self action:@selector(handleGetAccessible:)],
     [[FBRoute GET:@"/wda/element/:uuid/accessibilityContainer"] respondWithTarget:self action:@selector(handleGetIsAccessibilityContainer:)],
-#if !TARGET_OS_TV
+#if TARGET_OS_TV
+    [[FBRoute POST:@"/element/:uuid/focuse"] respondWithTarget:self action:@selector(handleFocuse:)],
+    [[FBRoute GET:@"/element/:uuid/focused"] respondWithTarget:self action:@selector(handleGetFocused:)],
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: do we need this endpoint?

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 5]
```diff
@@ -211,7 +219,38 @@ + (NSArray *)routes
   return FBResponseWithElementUUID(elementUUID);
 }
 
-#if !TARGET_OS_TV
+#if TARGET_OS_TV
++ (id<FBResponsePayload>)handleGetFocused:(FBRouteRequest *)request
+{
+  // `BOOL isFocused = [elementCache elementForUUID:request.parameters[@"uuid"]];`
+  // returns wrong true/false after moving focus by key up/down, for example.
+  // Thus, ensure the focus comparing the status with `fb_focusedElement`.
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: compares

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 6]
```diff
@@ -211,7 +219,38 @@ + (NSArray *)routes
...
+}
+
++ (id<FBResponsePayload>)handleFocuse:(FBRouteRequest *)request
+{
+  NSString *elementUUID = request.parameters[@"uuid"];
+  FBElementCache *elementCache = request.session.elementCache;
+  XCUIElement *element = [elementCache elementForUUID:elementUUID];
+  NSError *error;
+
+  if (![element fb_setFocusWithError:&error]) {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m:252]: what if element is not found in cache?

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 7]
```diff
@@ -67,7 +67,11 @@ + (NSArray *)routes
     [[FBRoute GET:@"/screenshot/:uuid"] respondWithTarget:self action:@selector(handleElementScreenshot:)],
     [[FBRoute GET:@"/wda/element/:uuid/accessible"] respondWithTarget:self action:@selector(handleGetAccessible:)],
     [[FBRoute GET:@"/wda/element/:uuid/accessibilityContainer"] respondWithTarget:self action:@selector(handleGetIsAccessibilityContainer:)],
-#if !TARGET_OS_TV
+#if TARGET_OS_TV
+    [[FBRoute POST:@"/element/:uuid/focuse"] respondWithTarget:self action:@selector(handleFocuse:)],
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: these are not standard endpoints, so it makes sense to prefix them all with `wda` (except of `attribute/focused`)

File: WebDriverAgentLib/Commands/FBFindElementCommands.m [hunk 8]
```diff
@@ -47,6 +47,9 @@ + (NSArray *)routes
     [[FBRoute POST:@"/element/:uuid/element"] respondWithTarget:self action:@selector(handleFindSubElement:)],
     [[FBRoute POST:@"/element/:uuid/elements"] respondWithTarget:self action:@selector(handleFindSubElements:)],
     [[FBRoute GET:@"/wda/element/:uuid/getVisibleCells"] respondWithTarget:self action:@selector(handleFindVisibleCells:)],
+#if TARGET_OS_TV
+    [[FBRoute GET:@"/wda/element/focused"] respondWithTarget:self action:@selector(handleFindFocusedElement:)],
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBFindElementCommands.m]: this looks like a duplicate of the endpoint above

File: WebDriverAgentLib/Utilities/FBTVNavigationTracker.m [hunk 9]
```diff
@@ -93,9 +93,10 @@ - (FBTVDirection)directionToFocusedElement
 }
 
 #pragma mark - Utilities
-- (id<FBElement>)focusedElement
+- (XCUIElement *)focusedElement
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Utilities/FBTVNavigationTracker.m]: this method looks very similar to `fb_focusedElement`

File: WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m (hunk 1 above)
Comment (KazuCocoa) [reply WebDriverAgentLib/Categories/XCUIElement+FBTVFocuse.m]: When I do `po self.wdEnabled` on debugging, no property error raised in XCUIElement. So just changed to `isEnabled`. (backed to wdEnabled. It also worked)

File: WebDriverAgentLib/Commands/FBFindElementCommands.m (hunk 8 above)
Comment (KazuCocoa) [reply WebDriverAgentLib/Commands/FBFindElementCommands.m]: Here gets a focused element in the display. So, it is similar to `/element/active`.

**Comment by KazuCocoa** [comment]
added New endpoints section in the description

File: Inspector/js/inspector.js [hunk 10]
```diff
@@ -56,6 +56,7 @@ class Inspector extends React.Component {
         {this.renderField('Rect', attributes.rect)}
         {this.renderField('isEnabled', boolToString(attributes.isEnabled))}
         {this.renderField('isVisible', boolToString(attributes.isVisible))}
+        {this.renderField('isFocused', boolToString(attributes.isFocused))}
```
Comment (mykola-mokhnach) [inline Inspector/js/inspector.js]: does it still work with iOS? should we have some condition there?

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 11]
```diff
@@ -211,7 +218,43 @@ + (NSArray *)routes
...
 }
 
-#if !TARGET_OS_TV
+#if TARGET_OS_TV
++ (id<FBResponsePayload>)handleGetFocused:(FBRouteRequest *)request
+{
+  // `BOOL isFocused = [elementCache elementForUUID:request.parameters[@"uuid"]];`
+  // returns wrong true/false after moving focus by key up/down, for example.
+  // Thus, ensure the focus compares the status with `fb_focusedElement`.
+  BOOL isFocused = false;
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: we usually use YES/NO constants in WDA

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 12]
```diff
@@ -211,7 +218,43 @@ + (NSArray *)routes
...
+
++ (id<FBResponsePayload>)handleFocuse:(FBRouteRequest *)request
+{
+  NSString *elementUUID = request.parameters[@"uuid"];
+  FBElementCache *elementCache = request.session.elementCache;
+  XCUIElement *element = [elementCache elementForUUID:elementUUID];
+  NSError *error;
+
+  if (!element) {
+    return FBResponseWithErrorFormat(@"'%@' element uuid didn't match any elements. Try find element again.",
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: ...the element again

File: WebDriverAgentLib/Commands/FBFindElementCommands.m [hunk 13]
```diff
@@ -117,6 +120,16 @@ + (NSArray *)routes
   return FBResponseWithCachedElement(element, request.session.elementCache, FBConfiguration.shouldUseCompactResponses);
 }
 
+#if TARGET_OS_TV
++ (id<FBResponsePayload>)handleGetFocusedElement:(FBRouteRequest *)request
+{
+  XCUIElement *element = request.session.activeApplication.fb_focusedElement;
+  if (nil == element) {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBFindElementCommands.m]: a ternary operator might looks better here

File: Inspector/js/inspector.js (hunk 10 above)
Comment (KazuCocoa) [reply Inspector/js/inspector.js]: In iOS case, it always shows `No`, but it does not need for them. Changed to return `nil` (the section disappears) if the attribute does not exist.

File: Inspector/js/inspector.js [hunk 14]
```diff
@@ -56,7 +56,7 @@ class Inspector extends React.Component {
         {this.renderField('Rect', attributes.rect)}
         {this.renderField('isEnabled', boolToString(attributes.isEnabled))}
         {this.renderField('isVisible', boolToString(attributes.isVisible))}
-        {this.renderField('isFocused', boolToString(attributes.isFocused))}
+        {this.renderField('isFocused', attributes.isFocused === undefined ? null : boolToString(attributes.isFocused))}
```
Comment (mykola-mokhnach) [inline Inspector/js/inspector.js]: typeof attributes.isFocused === 'undefined'

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by KazuCocoa** [comment]
@shvul 
Hi, I would share this with you.
Your FB's tvOS support PR has been merged into appium fork with some modifications.
Thanks so much.
https://github.com/appium/appium-xcuitest-driver/pull/911

**Comment by shvul** [comment]
@KazuCocoa 
Wow! That's amazing! 
Thanks a lot!

**Pull request merged**